### DISCUSSION TITLE
5.4.1 patch release this week?

### DISCUSSION THREAD

**Discussion by joelostblom** [vega/altair#3548] [General]
Would everyone be OK if I make a 5.4.1 patch release in the next few days? We're freezing our docker images to use for the semester and I would like to have https://github.com/vega/altair/pull/3533 included as I think that would be helpful for students. There are no Vega-Lite updates since the last release, so this release would be simple.

As per our recently more explicit effort to align with semantic versioning https://github.com/vega/altair/pull/3488 we should ideally not include [any commits marked with `feat`](https://github.com/vega/altair/commits/main/). Based on this, I'm thinking I will tag cfab6601aa649227c20880d05ae27799297694c3 a s 5.4.1 so the only two commits are the one I mentioned above and a doc update. If anyone feels strongly about tagging another commit as 5.4.1, let me know.

**Comment by binste** [comment]
Works for me, thank you Joel! :) Good to get #3533 released.

**Reply by MarcoGorelli** [reply]
Cool, thanks!

I also wanted to flag https://github.com/marimo-team/marimo/pull/2006 to you. It's the only case I've noticed of someone picking up on the change - they do say "This is nice actually nice since now we can pass polars, pyarrow, etc to altair charts and get back the polars, pyarrow, etc charts.", but also comment that it's technically a breaking change for people registering their own data transformers

I think this may be worth addressing? I'm just looking into it now, sorry for not having got round to it sooner

EDIT: https://github.com/vega/altair/issues/3549 looks "real" - addressing this now

**Reply by joelostblom** [reply]
Thanks for fixing that issue, let's wait for it to land as well before releasing then. After that PR is merged, we will have a `fix` commit after two `feat` commits on main, so now we can do one of the following:

1. Release the latest commit on main as 5.4.1 although there technically are two `feat` commits there.
2. Release the latest commit on main as 5.5.0 to more strictly adhere to SemVer.
3. Rebase main to move the fixes and version bumps before the `feat` commits (what I would have done on a smaller personal project, but maybe not a good idea here).

I'm leaning towards (1), because there are no notable user-facing features added in this release and it is more of a patch with bug fixes, although there technically are `feat` commits included, but I'm totally fine if we want to do (2) instead. What do you all think?

**Reply by dangotbanned** [reply]
@joelostblom PR is now merged

- #3549

**Comment by dangotbanned** [comment]
Thanks @joelostblom for the thoughtful usage of our [Versioning](https://altair-viz.github.io/about/versioning.html) policy!

# General
I would like to highlight that I marked https://github.com/vega/altair/pull/3535 and https://github.com/vega/altair/pull/3538 as `feat` due to changes in code generation, not for [functionality](https://altair-viz.github.io/about/versioning.html#version-numbers).

The usage of PR titles wasn't something I anticipated in the policy. Technically, they are based on a *related* standard [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
I can't speak for everyone, but I personally may have used different titles if I referred to [Why Use Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#why-use-conventional-commits) each time.
I use https://github.com/vivaxy/vscode-conventional-commits on commits and usually am just picking which description fits the closest

Is this worth opening another issue to discuss refining the policy language?

# PRs to consider
- https://github.com/vega/altair/pull/3530
	- No public API changes, I think would be beneficial to UX but have been waiting on review for 2 weeks
- https://github.com/vega/altair/pull/3543
	- No public API changes, possible perf improvements
- https://github.com/vega/altair/pull/3528
	- Easy include
- https://github.com/vega/altair/pull/3545
	- Wouldn't say I feel strongly, but is another fix.

# CI
These aren't user-facing, however there may be issues in CI with/without some/all of them:
- https://github.com/vega/altair/pull/3527
- https://github.com/vega/altair/pull/3529
- https://github.com/vega/altair/pull/3531
- https://github.com/vega/altair/pull/3539
- https://github.com/vega/altair/pull/3541

# Possible Bugs
I'm expecting this to cause an issue for anyone running `python>=3.13`, originally thought it was the source of https://github.com/pola-rs/polars/pull/17995#issuecomment-2282801814:
https://github.com/vega/altair/blob/591bf40fc0158ddd8d2fad2774d5230eb73ab734/pyproject.toml#L14-L18

Possibly an issue, since we have `typing_extensions` imports **outside** of conditional blocks.

## Fine
https://github.com/vega/altair/blob/591bf40fc0158ddd8d2fad2774d5230eb73ab734/altair/expr/__init__.py#L10-L13

## Will raise `ImportError`
https://github.com/vega/altair/blob/591bf40fc0158ddd8d2fad2774d5230eb73ab734/altair/expr/core.py#L3-L4

## Solution(s)
1. Replace the dependency with `"typing_extensions>=4.10.0"`
2. Rewrite all of the offending imports
3. Do nothing, as we have not declared support for `python>=3.13`

**Reply by joelostblom** [reply]
Thanks for the detailed response @dangotbanned!

> I would like to highlight that I marked https://github.com/vega/altair/pull/3535 and https://github.com/vega/altair/pull/3538 as feat due to changes in code generation, not for [functionality](https://altair-viz.github.io/about/versioning.html#version-numbers).

That sounds similar to my rationale in https://github.com/vega/altair/discussions/3548#discussioncomment-10405051, so it seems reasonable to release 5.4.1 from the latest commit on main after https://github.com/vega/altair/pull/3550 is merged.

> I use https://github.com/vivaxy/vscode-conventional-commits on commits and usually am just picking which description fits the closest... Is this worth opening another issue to discuss refining the policy language?

I think it is totally ok that we have some leeway in the interpretation of what we consider new functionality. And like you said, each commit that is a feature does not necessarily imply new functionality. I think I was a bit strict in the interpretation there; there is not necessarily a 1 to 1 mapping to the PR titles.

---

As for the Python 3.13 bugs, I think it is reasonable that we fix those once we officially declare support for 3.13.

**Reply by dangotbanned** [reply]
Sounds good to me. 

For future reference, I do want to voice that I'm in favour of more frequent, smaller **MINOR** releases. 
Even if only a small user group would benefit from the new functionality

**Comment by dangotbanned** [comment]
@joelostblom could we hold off on the patch release until these are resolved please?

- https://github.com/vega/altair/issues/3552
- #3553

**Reply by joelostblom** [reply]
Yes, sure! I will be on a plane most of tomorrow, so don't want to release until I land at the earlier (but the next few days are also fine from my end), just so I don't mess something up and then is not available to fix issues.

**Reply by dangotbanned** [reply]
Fine by me @joelostblom, I'm not in a rush so ping me if you have a deadline to meet for the docker stuff

**Reply by joelostblom** [reply]
I just checked in with your IT staff and if there is a new version up on Tue morning, then it would be good from their side. So if I can release 5.4.1 Mon night, that would be great as it takes a few hours usually to propagate to conda which is what we use in our Docker images. I think #3553 can be merged by then and even if #3552 is not ready yet (having the fixes up until now as 5.4.1 would be beneficial for our docker images, and if #3552 lands after that then altair can still follow up with 5.4.2 soon after even if that doesn't make it into our images; as you said, frequent smaller bug fixes releases are welcome)

**Reply by dangotbanned** [reply]
All good!
I'll likely put something together later today/tomorrow for #3552 if @binste doesn't beat me to it

**Reply by dangotbanned** [reply]
@joelostblom just to keep you updated, I'm thinking it might be best to hold off on resolving #3552 for now.

I proposed some alternative solutions in https://github.com/vega/altair/issues/3552#issuecomment-2307828887 which I feel are more beneficial than the pragmatic one that @binste had already agreed to in https://github.com/vega/altair/issues/3552#issuecomment-2305351048

Would feel better merging one of those, but wouldn't want to push it through without approval.

Feel free to go ahead with `v5.4.1` release!

**Reply by joelostblom** [reply]
Thanks for the update! I will tag e543fbc95c40de9f99f109100dec50bc43b9d5db as 5.4.1 then and go ahead and create the release in a few hours.

**Comment by joelostblom** [comment]
[Released](https://github.com/vega/altair/releases/tag/v5.4.1), thanks everyone!

**Reply by binste** [reply]
Thanks @joelostblom 🥳

**Comment by dangotbanned** [comment]
@joelostblom @binste 

Just noticed that since `v5.4.1` the typing docs are showing the same as what I found in https://github.com/vega/altair/pull/3515#discussion_r1705732684

- https://altair-viz.github.io/user_guide/api.html#typing
- https://altair-viz.github.io/user_guide/generated/typing/altair.typing.ChannelDescription.html#altair.typing.ChannelDescription
- https://altair-viz.github.io/user_guide/generated/typing/altair.typing.ChannelDetail.html#altair.typing.ChannelDetail

**Reply by joelostblom** [reply]
Is this fixed on main? (sorry on my phone and can't check right now). I was thinking of pushing an update to the docs after https://github.com/vega/altair/pull/3544 is merged, even if there is not a new release by then since that PR documents important functionality of 5.4.1.

**Reply by dangotbanned** [reply]
@joelostblom sort of

We [settled on a temporary solution](https://github.com/vega/altair/pull/3515#discussion_r1710083025) to build the docs using `3.11` - so you might be able to rebuild and get the same result as `v5.4.0` that way?

I still don't fully understand why that worked.

---

FYI I'm not viewing this as an urgent problem (would have raised an issue), just wanted to note it since it wasn't what I expected to see
